### ISSUE TITLE
npm 5.0.0/5.0.1: npm update does not install missing dev dependencies

### ISSUE THREAD

**Issue by gurpreetatwal** [npm/npm#16901]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [x] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

According to the [docs](https://docs.npmjs.com/cli/update), `npm update` should "also install missing packages"  however both npm 5 and npm 5.0.1 do not install missing dev dependencies if the package has no normal dependencies. I have tried with and without a `package-lock.json` in the folder as well as with and without the `--dev` flag.

#### How can the CLI team reproduce the problem?

```bash
npm init
npm install is-regexp --save-dev
rm -rf node_modules
npm update 
npm update --dev
ls 
```
npm-debug.log
```
npm info it worked if it ends with ok
npm verb cli [ '/usr/local/bin/node', '/usr/local/bin/npm', 'update' ]
npm info using npm@5.0.1
npm info using node@v6.9.2
npm verb exit [ 0, true ]
npm info ok
```

### supporting information:

 - `npm -v` prints:  `5.0.1`
 - `node -v` prints: `6.9.2`
 - `npm config get registry` prints: `https://registry.npmjs.org/`
 - Windows, OS X/macOS, or Linux?:  `Linux Mint 18 Sarah`

**Comment by gurpreetatwal** [comment]
Its seems that this only happens when the package *ONLY* has dev dependencies. If a package has normal dependencies then both the dev and normal dependencies are installed.

So the following example does produce the expected behavior
```
npm init
npm install avsc
npm install is-regexp --save-dev
rm -rf node_modules
npm update 
ls 
```

**Comment by mitsos1os** [comment]
@gurpreetatwal What you say is different than my issue. You do not mention any local folder module install. Also I tried reproducting your steps and they work correctly. Both `avsc` and `is-regexp` are re-installed after `npm update`

**Comment by gurpreetatwal** [comment]
@mitsos1os I think you might be responding to the wrong issue? I haven't talked to anyone about this issue other than my coworkers. 

Also I believe you ran the steps that I had in my follow up comment (the ones that installs both `avsc` and `is-regexp`), if you read the sentence right above it it states: *the following example does produce the expected behavior*.

The reproduction steps for the bug are in my original comment.

**Comment by mitsos1os** [comment]
@gurpreetatwal , I am really sorry. I had the tab open... and accidentally saw your reply and though it was on an other topic... Too tired I guess... Would you like me to delete the comment?

**Comment by gurpreetatwal** [comment]
@mitsos1os  haha no worries,  I was just really confused. In terms of deleting the comment, up to you.

Get some rest though! lol

**Comment by mitsos1os** [comment]
@gurpreetatwal Before resting....I will delete the comment to keep the conversation clear.... Just informing you so that you can delete yours as well to avoid having unrelated replies... I will also delete this last one as soon as you delete yours. Sorry again for the trouble!

**Comment by gurpreetatwal** [comment]
@mitsos1os gotcha! and no worries :smile:

**Comment by amiram** [comment]
@gurpreetatwal I can reproduce it even if I have also regular deps. Running `update` or `update --dev` makes no difference. It doesn't update my dev deps, and even deleting the dev packages, since npm thinks they are not needed any more.

**Comment by gurpreetatwal** [comment]
@amiram Hmm, I tried it with regular deps and it seems to be working fine for me. Can you post a simple reproduction? It could help the npm team find the underlying issue(s). 

FWIW Here's what I tried:

```
$ npm init
$ npm install avsc
$ npm install is-plain-object@2.0.0 --save-dev

+ is-plain-object@2.0.0
added 2 packages in 1.335s

# verify that package.json has is-plain-object: ^2.0.0 under devDependencies

$ npm update

+ is-plain-object@2.0.3
added 1 package, removed 1 package and updated 1 package in 0.87s


$ npm ls 
/home/gurpreet/temp/npm-update-dev
├── avsc@5.0.4
└─┬ is-plain-object@2.0.3
  └── isobject@3.0.0

```

**Comment by amiram** [comment]
I tried to repro it but no luck. When it happened it was in a big project, not a sample. I'll update if it happens again.

**Comment by jfstgermain** [comment]
I'm getting the same problem
```
$ npm -v
5.0.3
$ npm init
$ npm install avsc
$ npm install is-plain-object --save-dev
$ less package.json
  "dependencies": {
    "avsc": "~5.0.4"
  },
  "devDependencies": {
    "is-plain-object": "~2.0.3"
  }
$ rm -Rf node_modules
$ npm install --production
$ npm update --dev
$ ls node_modules 
avsc

```

**Comment by seanlindo** [comment]
Same here. Everything was working fine on my last version of NPM, now I can't reliably use my boilerplate projects without installing each package individually.

v5.0.3

**Comment by kloni** [comment]
My experience (v5.0.3): "npm install" does not install (dev)dependencies when just adding them to package.json and having a preexisting package-lock.json. I could work around by removing package-lock.json as suggested in  #16839.

**Comment by seanlindo** [comment]
I'm going to try that out @kloni

**Comment by lanwin** [comment]
I have exactly this problem since updating to node 8. I have both normal and dev deps but "npm install" forgets the dev deps. Even if I remove the lock file. Even if I do "npm install foo" it removes my dev deps. The only chance to get my dev deps is via "npm install --dev".

But they never seems to get it to the lock file, which seems to be the problem here.

**Comment by lanwin** [comment]
Hu I noticed that the dev config option is default to false. Does this mean the default behavior of "npm intsall" have changed to install only dependencies but not devdependencies anymore?

**Comment by dannielum** [comment]
I had this issue too, reinstalling node and npm seems to fix the issue. I believe you need npm v3.10.3 for node v6.x.

**Comment by lanwin** [comment]
@dannielum do you mean a downgrade?

**Comment by lanwin** [comment]
If I going to the NPM issue list I can see a lot what for me seems to be related to this. 

https://github.com/npm/npm/issues/17384
https://github.com/npm/npm/issues/17263
https://github.com/npm/npm/issues/17337
https://github.com/npm/npm/issues/17324
https://github.com/npm/npm/issues/17379
https://github.com/npm/npm/issues/17165

It seems that NPM5:
* Ignores dev dependencies for package-lock.json 
And at the same time 
* it forces its packages to meet the package-lock.json on every install (remove everything else)

Are the NPM devs aware of that?

**Comment by dannielum** [comment]
@lanwin, I uninstalled npm and then installed the older version of node v6.x which also installed with npm v3.x. That fixed my issue.

**Comment by gurpreetatwal** [comment]
@lanwin @dannielum npm 5 (and node 8 for that matter) is essentially the "bleeding edge" release and there are a few known issues with the release. [The developers are aware of these issues and are working to resolve them.](https://github.com/npm/npm/issues/16991#issuecomment-310827750)

If npm 5 is currently breaking your workflow you can downgrade to npm 4 by running `npm install -g npm@latest-4` and check back with npm 5 in a couple of weeks.

I made this issue specifically for the bug regarding `npm update` and dev dependencies.

**Comment by amiram** [comment]
Anyone saw this fixed in 5.1.0?

**Comment by amiram** [comment]
Still happening :(
Running npm update delete dev dependencies in some of my projects.

**Comment by Standard8** [comment]
@amiram The specific issue I had in #17384 is fixed, although that wasn't dealing with dev dependencies.

**Comment by amiram** [comment]
Tried with 5.2.0. Didn't happen yet.

**Comment by amiram** [comment]
Now it did. It is unpredictable.

**Comment by gurpreetatwal** [comment]
@amiram I had tried it with 5.1.0 and it didn't work for me. I'll give it a shot with 5.2.0 later today and report back


(FWIW the this issue isn't marked as done on #16991, but I doubt its a high priority for them to go mark things as done vs getting things done)

**Comment by gurpreetatwal** [comment]
doesn't work for me on npm 5.2.0 either

**Comment by gurpreetatwal** [comment]
or 5.3.0 for that matter

**Comment by ifuller1** [comment]
I'm still seeing this on Mac OS X too. Anything we can do to help progress this fix, any logs that would be helpful?

**Comment by amiram** [comment]
Anything with 5.4.0?

**Comment by gurpreetatwal** [comment]
Just tried it, no change

Maybe we'll get a fix with npm v6 :stuck_out_tongue:

**Comment by gurpreetatwal** [comment]
No idea if anyone is still following this, but still broken with v5.5.1 🎉

**Comment by amiram** [comment]
@gurpreetatwal node should start v8 LTS in 12 days. Is it going to be with this bug?

**Comment by gurpreetatwal** [comment]
lol, I'm assuming so; should be interesting

There was some discussion about reverting node 8 to use npm 4, but conclusion was that would be considered a breaking change and therefore would have to be tabled until node 9 

https://github.com/nodejs/node/issues/13815

**Comment by amiram** [comment]
How did you reproduced it? I tried with @jfstgermain script and indeed it happened, but sometimes it happened to me even without the --production flag. I just did npm update for several repos at once and randomly some of them lost the dev packages.

**Comment by gurpreetatwal** [comment]
To give credit where it's due, they have fixed a large number of bugs, but npm 5 still doesn't feel like a polished release.  It was becoming a daily routine at my company for someone to have to run `rm -rf node_modules package-lock.jdon` because npm 5 had somehow messed up their install . We had to go so far as add config to every repo to not create `package-lock.json` files.

**Comment by amiram** [comment]
Even if node 8 is shipped with npm 5, you can still use npm 4 if you want

**Comment by gurpreetatwal** [comment]
tl;dr use the `--save-dev`
```bash
npm update --save-dev
``` 

Decided to look into this issue a bit more since Node 8 went LTS and all of my Travis builds broke because I was relying on `npm update` and build caching.  After looking through the code, I'm not quite sure why this broke between version 4 and version 5 as the code that [handles checking is ~2 years old](https://github.com/npm/npm/blame/latest/lib/outdated.js#L228). However the code does rely on some config settings being set so maybe defaults changed? I'll try to took a look at this some more and see if I can find a fix that can actually be PR'd

**Comment by gurpreetatwal** [comment]
Yup, hunch turned out to be correct. [This commit](https://github.com/npm/npm/commit/c0d73dbed4f1e5699de1914c1cc291973ad87c9f#diff-c147c0dd597c04e90f7efaa203b5d0c7L196) changed the default value for `save` from `false` to `true` and in doing so caused this bug to rear its head.

Obviously,  this could be "fixed" by changing the default back, but this would probably lead to other headaches and issues since defaulting saves have become the accepted behavior for npm 5. The real solution is to refactor the checks in the `outdated.js` file.

**Comment by gurpreetatwal** [comment]
So depending on your expected behavior either of the following commands should update dev dependencies.

---

1. You want to save the new version of the dev deps
```
npm update --save-dev
```

2. no saving pls
```
npm update --save false
```

**Comment by jcrben** [comment]
`npm update <package> --save-dev` isn't doing anything for me (in this case, ts-loader ^3.0.0) with npm@5.5.1. 

Now that we're into locking by default, npm update's purpose should be able to update the `package-lock.json`. All that matters in package.json is the semver range. 

By default npm install will update `package-lock.json` to the latest in the acceptable range for me in npm@5.5.1 - which is not necessarily what I want and seems to be a bug (see https://github.com/npm/npm/issues/18286#issuecomment-327285191), but it is what is happening. Hopefully down the line `npm install` will default to the exact versions in the lockfile while `npm update` will update the lockfile according to the appropriate semver range.

To make this more clear, I might use asterisks instead of numbers (e.g. `ts-loader: '^3.*.*'`) altho I think most people use zeros for this purpose.

Some people might like the convenience of seeing the actual version in their package.json file rather than the package-lock.json, altho I think third-party tools like https://github.com/peerigon/updtr could handle that...
